### PR TITLE
feat(ethereum contracts): add emergency stop for relayer contract

### LINKED ISSUE #76: Emergency stop
If 2 proofs are submitted and both are valid but have conflicting data then bridge must be halted.

### PR DESCRIPTION
**Pull request by ByteNacked**
Resolves #76.

Add emergency stop flag for ethereum relayer contract which when activated halts the bridge message propagation (effectively halts the bridge).

@reviewer-or-team

### CODE REVIEW

File: ethereum/src/MessageQueue.sol [hunk 1]
```diff
@@ -52,6 +52,11 @@ contract MessageQueue is IMessageQueue {
         VaraMessage calldata message,
         bytes32[] calldata proof
     ) public {
+        if (IRelayer(RELAYER_ADDRESS).emergencyStop()) {
+            // Emergency stop is active, stop processing messages.
+            return;
+        }
```
Comment (mertwole) [inline ethereum/src/MessageQueue.sol:58]: I think it's better to revert inside `getMerkleRoot` to guarantee that `RelayerContract` is totally unuseful when emergency stop occured

File: ethereum/src/Relayer.sol [hunk 2]
```diff
@@ -105,4 +122,24 @@ contract Relayer is IRelayer {
...
     }
+
+    /**
+     * @dev Checks if the merkle root provided is duplicated or not.
+     * If it is duplicated, it sets the emergency mode.
+     *
+     * @param block_number Target block number.
+     * @param merkle_root Target merkle root.
+     */
+    function _checkEmergencyStopCondition(
```
Comment (mertwole) [inline ethereum/src/Relayer.sol:133]: I think it can be inlined and done without `if` condition

File: ethereum/src/interfaces/IRelayer.sol [hunk 3]
```diff
@@ -26,4 +27,5 @@ interface IRelayer {
         bytes32 merkle_root
     ) external pure returns (uint256[] memory public_inputs);
 
+    function emergencyStop() external view returns (bool);
```
Comment (mertwole) [inline ethereum/src/interfaces/IRelayer.sol:30]: Please also update `ethereum/client/src/abi.rs`

File: ethereum/src/Relayer.sol (hunk 2 above)
Comment (ByteNacked) [reply ethereum/src/Relayer.sol:133]: Are you suggest to inline this function?

Comment (mertwole) [reply ethereum/src/Relayer.sol:133]: yes

File: ethereum/src/interfaces/IRelayer.sol (hunk 3 above)
Comment (ByteNacked) [reply ethereum/src/interfaces/IRelayer.sol:30]: Fixed

File: ethereum/src/MessageQueue.sol (hunk 1 above)
Comment (ByteNacked) [reply ethereum/src/MessageQueue.sol]: Fixed

File: ethereum/test/MessageQueue.t.sol [hunk 4]
```diff
@@ -423,4 +423,18 @@ contract MessageQueueTest is TestHelper {
 
         assertEq(erc20_token.balanceOf(ETH_ADDRESS_3), 10 * (10 ** 18));
     }
+
+    function test_relayer_contract_emergency_mode() public {
```
[the remark this answers is not in the document]
Comment (ByteNacked) [reply ethereum/test/MessageQueue.t.sol:427]: Fixed

**Pull request merged**